### PR TITLE
Remove HTTPManager and dependent interfaces

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
The indirection is unnecessary and can introduce future versioning
issues.  Instead, create http.Request where needed and send them using
the configured HTTPClient.
Flow context parameter to HTTP request creation.
Use http.Header and url.Values types instead of map[string]string.
Use query parameter encoding algorithm from the standard library.
Misc other bits of refactoring.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/56

**Comment by jhendrixMSFT** [comment]
One test failure that needs to be resolved.

File: msal/mock_http_manager.go [hunk 1]
```diff
@@ -3,20 +3,20 @@
...
-func (mock *mockHTTPManager) Get(url string, requestHeaders map[string]string) (HTTPManagerResponse, error) {
-	args := mock.Called(url, requestHeaders)
-	return args.Get(0).(HTTPManagerResponse), args.Error(1)
-}
-
-//Post mocks the Post method of a HTTPManager
-func (mock *mockHTTPManager) Post(url string, body string, requestHeaders map[string]string) (HTTPManagerResponse, error) {
-	args := mock.Called(url, body, requestHeaders)
-	return args.Get(0).(HTTPManagerResponse), args.Error(1)
+func (m *mockHTTPManager) Do(req *http.Request) (*http.Response, error) {
```
Comment (sangonzal) [inline msal/mock_http_manager.go:16]: Can this file be removed given that `http_manager` is being removed?

**Review by sangonzal — APPROVED** [review]
Looks good to me! I think the new http code looks a lot cleaner. There are some future features (like telemetry and logging) where having a single point that all http requests and responses go through (formerly msal_http_manager.go) is useful, but that can be done later on if it's needed.

File: msal/mock_http_manager.go (hunk 1 above)
Comment (jhendrixMSFT) [reply msal/mock_http_manager.go:16]: Not at present as this is still used as part of the mock testing.

File: internal/requests/device_code_response.go [hunk 2]
```diff
@@ -28,14 +30,19 @@ type DeviceCodeResponse struct {
 
 // CreateDeviceCodeResponse creates a deviceCodeResponse instance from HTTP response.
 // TODO(jdoak): look at responseData and see if we can get it originally from []byte instead.
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply internal/requests/device_code_response.go]: Good catch, fixed.

**Review by keegan-caruso — APPROVED** [review]
LGTM

File: internal/msalbase/oauth_response_base_test.go [hunk 3]
```diff
@@ -9,20 +9,20 @@ import (
...
 func TestCreateOAuthResponseBase(t *testing.T) {
-	_, err := CreateOAuthResponseBase(404, oauthResponse)
+	_, err := CreateOAuthResponseBase(404, []byte(oauthResponse))
 	actualError := errors.New("HTTP 404")
 	if err.Error() != actualError.Error() {
 		t.Errorf("Actual error %v differs from expected error %v", err, actualError)
 	}
-	expectedErr := errors.New("invalid request")
-	_, err = CreateOAuthResponseBase(300, oauthResponseWithError)
+	expectedErr := errors.New("invalid request: missing payload content")
```
Comment (element-of-surprise) [inline internal/msalbase/oauth_response_base_test.go:20]: These expectedErr are brittle.  Any change to the text of the error can cause the tests to fail.

I know that wasn't your test, but can we change this so we just test the given inputs against outputs?  The output can be simply tested to see if it has an error, not a specific error.  This is sufficient 99.9% of the time.

File: internal/requests/auth_code_request.go [hunk 4]
```diff
@@ -38,28 +40,28 @@ func CreateAuthCodeRequest(webRequestManager WebRequestManager, authParameters m
...
 	resolutionManager := CreateAuthorityEndpointResolutionManager(req.webRequestManager)
-	endpoints, err := resolutionManager.ResolveEndpoints(req.authParameters.AuthorityInfo, "")
+	endpoints, err := resolutionManager.ResolveEndpoints(ctx, req.authParameters.AuthorityInfo, "")
 	if err != nil {
 		return msalbase.TokenResponse{}, fmt.Errorf("unable to resolve endpoints: %w", err)
 	}
 
 	req.authParameters.Endpoints = endpoints
-	params := make(map[string]string)
+	params := url.Values{}
```
Comment (element-of-surprise) [inline internal/requests/auth_code_request.go:51]: Thank you, this has been tickling my ocd.

File: internal/wstrust/wstrust_mex_document.go [hunk 5]
```diff
@@ -49,9 +51,14 @@ func updateEndpoint(cached *Endpoint, found Endpoint) bool {
 
 // TODO(jdoak): Refactor into smaller bits
 
-func CreateWsTrustMexDocument(responseData string) (MexDocument, error) {
+func CreateWsTrustMexDocument(resp *http.Response) (MexDocument, error) {
 	definitions := &definitions{}
```
Comment (element-of-surprise) [inline internal/wstrust/wstrust_mex_document.go:55]: Nit: I'd move definitions down to right above line 61 where it is first used.

**Review by element-of-surprise — CHANGES_REQUESTED** [review]

File: internal/msalbase/oauth_response_base_test.go (hunk 3 above)
Comment (jhendrixMSFT) [reply internal/msalbase/oauth_response_base_test.go]: Agreed, fixed.

File: internal/wstrust/wstrust_mex_document.go (hunk 5 above)
Comment (jhendrixMSFT) [reply internal/wstrust/wstrust_mex_document.go]: Fixed.

**Pull request merged**